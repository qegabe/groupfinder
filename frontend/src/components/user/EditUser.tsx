import React, { useEffect, useState } from "react";
import { Box, Button, TextField, Typography } from "@mui/material";
import useFormData from "../../hooks/useFormData";
import GroupFinderApi from "../../api";
import { useAppDispatch, useAppSelector } from "../../hooks/redux";
import parseFormErrors from "../../helpers/parseFormErrors";
import { Link } from "react-router-dom";
import { setAlert, setAvatar } from "../../actions/actionCreators";

function EditUser() {
  const [formErrors, setFormErrors] = useState<any>({});
  const [formData, handleChange, setFormData] = useFormData({
    avatarUrl: "",
    bio: "",
  });
  const user = useAppSelector((s) => s.auth.user);
  const dispatch = useAppDispatch();

  useEffect(() => {
    async function loadUser() {
      const userData = await GroupFinderApi.getUser(user?.username as string);
      setFormData((fd) => ({
        avatarUrl: userData.avatarUrl || "",
        bio: userData.bio || "",
      }));
    }
    loadUser();
  }, [user?.username, setFormData]);

  async function handleSubmit(evt: any) {
    evt.preventDefault();
    try {
      await GroupFinderApi.updateProfile(user?.username as string, formData);
      dispatch(setAvatar(formData.avatarUrl));
      dispatch(setAlert("success", "Saved"));
    } catch (error: any) {
      if (typeof error === "string") {
        setFormErrors(parseFormErrors(error));
      } else console.error(error);
    }
  }

  return (
    <Box sx={{ display: "grid", justifyItems: "center" }}>
      <Typography variant="h3" sx={{ mt: 2 }}>
        Edit Profile
      </Typography>
      <Box
        component="form"
        autoComplete="off"
        sx={{
          display: "grid",
          my: 3,
          px: 1,
          width: { xs: "100%", lg: "60%" },
          justifyItems: "center",
        }}
        onSubmit={handleSubmit}>
        <TextField
          fullWidth
          id="avatarUrl"
          label="Avatar URL"
          variant="outlined"
          type="url"
          sx={{ my: 1 }}
          value={formData.avatarUrl}
          onChange={handleChange}
          {...formErrors.avatarUrl}
        />
        <TextField
          fullWidth
          id="bio"
          label="Bio"
          multiline
          maxRows={4}
          variant="outlined"
          sx={{ my: 1 }}
          value={formData.bio}
          onChange={handleChange}
          {...formErrors.bio}
        />
        <Box
          sx={{
            width: "100%",
            display: "flex",
            justifyContent: "center",
            my: 2,
          }}>
          <Button type="submit" variant="contained" sx={{ mr: 1 }}>
            Save
          </Button>
          <Button
            component={Link}
            to={user ? `/users/${user.username}` : "/"}
            variant="outlined"
            color="secondary">
            Cancel
          </Button>
        </Box>
      </Box>
    </Box>
  );
}

export default EditUser;
