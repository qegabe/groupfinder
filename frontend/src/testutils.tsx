import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

function customRender(
  ui: ReactElement,
  route?: string,
  preloadedState?: PreloadedState<RootState>,
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={route ? [route] : undefined}>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {children}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
