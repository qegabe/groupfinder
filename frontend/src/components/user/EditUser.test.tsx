import React from "react";
import { render, waitFor, screen, fireEvent } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import EditUser from "./EditUser";

const mockGetUser = jest
  .spyOn(GroupFinderApi, "getUser")
  .mockImplementation(async () => {
    return {
      username: "TestUser",
      bio: "Test Bio",
      avatarUrl: null,
      triviaScore: 1000,
    };
  });

const mockUpdateProfile = jest
  .spyOn(GroupFinderApi, "updateProfile")
  .mockImplementation(async () => {});

afterEach(() => {
  jest.clearAllMocks();
});

it("renders", async () => {
  const { container } = render(<EditUser />);
  await waitFor(() => {
    const bioInput = container.querySelector("#bio");
    expect(bioInput).toHaveValue("Test Bio");
  });
  expect(mockGetUser).toHaveBeenCalledTimes(1);
});

it("matches snapshot", async () => {
  const { container, asFragment } = render(<EditUser />);
  await waitFor(() => {
    const bioInput = container.querySelector("#bio");
    expect(bioInput).toHaveValue("Test Bio");
  });
  expect(asFragment()).toMatchSnapshot();
});

it("lets user update profile", async () => {
  const { container } = render(<EditUser />, undefined, {
    auth: {
      token: null,
      user: { username: "TestUser", avatarUrl: null },
      error: null,
    },
  });
  await waitFor(() => {
    const bioInput = container.querySelector("#bio");
    expect(bioInput).toHaveValue("Test Bio");
  });
  const avatarUrlInput = container.querySelector("#avatarUrl");
  const saveButton = screen.getByText("Save");
  fireEvent.change(avatarUrlInput as Element, {
    target: { value: "http://some.img" },
  });
  fireEvent.click(saveButton);
  await waitFor(() => {
    expect(screen.getByText("Saved")).toBeInTheDocument();
  });
  expect(mockUpdateProfile).toHaveBeenCalledWith("TestUser", {
    bio: "Test Bio",
    avatarUrl: "http://some.img",
  });
});
