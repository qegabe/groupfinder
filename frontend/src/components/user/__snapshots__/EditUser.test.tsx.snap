// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1tx56hj"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
    >
      Edit Profile
    </h3>
    <form
      autocomplete="off"
      class="MuiBox-root css-6feot8"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="avatarUrl"
          id="avatarUrl-label"
        >
          Avatar URL
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="avatarUrl"
            type="url"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Avatar URL
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="bio"
          id="bio-label"
        >
          Bio
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id="bio"
            style="height: 0px; overflow: hidden;"
          >
            Test Bio
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Bio
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiBox-root css-owjlnq"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Save
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          href="/"
          tabindex="0"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
